eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/OfferCodecs.scala

```diff
@@ -0,0 +1,227 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fr.acinq.eclair.wire.protocol
+
+import fr.acinq.bitcoin.scalacompat.Crypto.PublicKey
+import fr.acinq.bitcoin.scalacompat.ByteVector32
+import fr.acinq.bitcoin.Bech32
+import fr.acinq.eclair.crypto.Sphinx.RouteBlinding.{BlindedNode, BlindedRoute}
+import fr.acinq.eclair.payment.Bolt12Invoice
+import fr.acinq.eclair.wire.protocol.CommonCodecs._
+import fr.acinq.eclair.wire.protocol.Offers._
+import fr.acinq.eclair.wire.protocol.OnionRoutingCodecs.MissingRequiredTlv
+import fr.acinq.eclair.wire.protocol.TlvCodecs.{tmillisatoshi, tu32, tu64overflow}
+import fr.acinq.eclair.{CltvExpiryDelta, Feature, Features, MilliSatoshi, TimestampSecond, UInt64}
+import scodec.bits.ByteVector
+import scodec.codecs._
+import scodec.{Attempt, Codec}
+
+import scala.util.Try
+
+object OfferCodecs {
+  private val chains: Codec[Chains] = variableSizeBytesLong(varintoverflow, list(bytes32)).xmap[Seq[ByteVector32]](_.toSeq, _.toList).as[Chains]
+
+  private val currency: Codec[Currency] = variableSizeBytesLong(varintoverflow, utf8).as[Currency]
+
+  private val amount: Codec[Amount] = variableSizeBytesLong(varintoverflow, tmillisatoshi).as[Amount]
+
+  private val description: Codec[Description] = variableSizeBytesLong(varintoverflow, utf8).as[Description]
+
+  private val features: Codec[FeaturesTlv] = variableSizeBytesLong(varintoverflow, bytes).xmap[Features[Feature]](Features(_), _.toByteVector).as[FeaturesTlv]
+
+  private val absoluteExpiry: Codec[AbsoluteExpiry] = variableSizeBytesLong(varintoverflow, tu64overflow).as[TimestampSecond].as[AbsoluteExpiry]
+
+  private val blindedNodeCodec: Codec[BlindedNode] = (("nodeId" | publicKey) :: ("encryptedData" | variableSizeBytes(uint16, bytes))).as[BlindedNode]
+
+  private val pathCodec: Codec[BlindedRoute] = (("firstNodeId" | publicKey) :: ("blinding" | publicKey) :: ("path" | listOfN(uint8, blindedNodeCodec).xmap[Seq[BlindedNode]](_.toSeq, _.toList))).as[BlindedRoute]
+
+  private val paths: Codec[Paths] = variableSizeBytesLong(varintoverflow, list(pathCodec)).xmap[Seq[BlindedRoute]](_.toSeq, _.toList).as[Paths]
+
+  private val issuer: Codec[Issuer] = variableSizeBytesLong(varintoverflow, utf8).as[Issuer]
+
+  private val quantityMin: Codec[QuantityMin] = variableSizeBytesLong(varintoverflow, tu64overflow).as[QuantityMin]
+
+  private val quantityMax: Codec[QuantityMax] = variableSizeBytesLong(varintoverflow, tu64overflow).as[QuantityMax]
+
+  // Only the x component of the key is serialized. We arbitrarily add hex"02" to get a valid key.
+  // When using this key, we need to try both the version starting with hex"02" and the one starting with hex"03".
+  private val nodeId: Codec[NodeId] = variableSizeBytesLong(varintoverflow, bytes32).xmap[PublicKey](b32 => PublicKey(2.toByte +: b32), xOnlyPublicKey).as[NodeId]
```

Comment:
You didn't like to have `NodeId` defined this way (in `Offers.scala`)?

```
case class NodeId(xonly: ByteVector32) extends OfferTlv with InvoiceTlv {
    val nodeId1: PublicKey = PublicKey(2 +: xonly)
    val nodeId2: PublicKey = PublicKey(3 +: xonly)
}
```

I find this better, it clearly encodes the subtlety while giving easy access to everything you would need.

Reply:
No I didn't like it because for invoices we may know which one it is (because we sent the invoice request to both and only one sent a response), so this type is no longer appropriate.

Reply:
6a0dcfbf5 uses your TLV and adds an explicit nodeId for the invoice when we know it, what do you think?

Reply:
Looks good, thanks!